chain/pool/src/lib.rs

```diff
@@ -118,6 +109,122 @@ impl TransactionPool {
     }
 }
 
+/// Draining Iterator is a structure to pull transactions from the pool.
+/// It allows to request a next transaction either with the new key (next key) or with the same key.
+/// When a draining iterator is dropped the remaining transactions are returned back to the pool.
+pub struct DrainingIterator<'a> {
+    /// Mutable reference to the pool, to avoid exposing it while the iterator exists.
+    pool: &'a mut TransactionPool,
+    /// Whether every group of transactions is already sorted.
+    sorted: bool,
+    /// Helper flag to know which map is the current one and which map is the next one.
+    /// Instead of swapping memory between maps, the implementation uses helper functions.
+    current_map_is_pool: bool,
+    /// The temporary map to hold transactions for which keys were already used in the current batch.
+    temp_map: TxMap,
+    /// The last entry from the current map. If it's None, it means the entry for the key was fully
+    /// drained or was not yet initialized.
+    /// If it's Some, then it guarantees that there are a least one transaction in the vector.
+    last_entry: Option<((AccountId, PublicKey), Vec<SignedTransaction>)>,
+}
+
+/// The iterator works with the following algorithm:
+/// 1. Initializes the current map to be the one from the pool.
+/// 2. An entry is pulled from the current map.
+///    2.1. If the current map is empty, swaps it with the next map.
+///    2.2. Remembers that all entries are sorted now.
+/// 3. If a not sorted yet, sorts the transactions in the entry in non-decreasing order by nonce, so
+///    a transaction with the lowest nonce is the last element.
+/// 4. If a new entry is needed for a new key, inserts the current entry to the next map.
+/// 5. Pulls the latest the transaction from the current entry.
+/// 6. If the current entry becomes empty, sets it to None.  
+impl<'a> DrainingIterator<'a> {
```

Comment:
I suggest we separate DrainingIterator into two iterators: one that iterates over the groups and another one that iterates over the elements within a group. Then both of them can actually implement `Iterator` trait.

Then instead of passing a boolean on each call of `next` we can have two nested loops: the outer loop iterates over the groups while the inner iterates over the transactions within the given group, and it only makes a step when the previous transaction was invalid.